src/lib/libast/tests/path/pathcanon.c

```diff
@@ -0,0 +1,99 @@
+#include "config_ast.h"  // IWYU pragma: keep
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "ast.h"
+#include "terror.h"
+
+const char *linkname = "foo";
+const char *filename = "foobarbaz";
+
+void cleanup_handler(void) {
+    unlink(linkname);
+    unlink(filename);
+}
+
+tmain() {
+    UNUSED(argc);
+    UNUSED(argv);
+
+    char path[PATH_MAX];
+    int fd;
+
+    atexit(cleanup_handler);
+
+    fd = open(filename, O_CREAT, 0666);
+    if (fd == -1) terror("Failed to create test file");
+    close(fd);
+
+    if (symlink(filename, linkname) < 0) {
+        terror("Failed to create symbolic link");
+    }
+
+    // Try to resolve physical path from symbolic link
+    strcpy(path, linkname);
+    pathcanon(path, PATH_MAX, PATH_PHYSICAL);
+    if (strcmp(path, filename)) terror("Failed to resolve symbolic link");
+
+    // Try to resolve physical path from broken symbolic link
+    unlink(linkname);
+    strcpy(path, linkname);
+    pathcanon(path, PATH_MAX, PATH_PHYSICAL | PATH_EXISTS);
+    fprintf(stderr, "%s\n", path);
+    if (strcmp(path, "")) terror("Resolving broken link should fail");
+
+    // TODO: Three dots. It is special cased in `pathcanon()` function. Why ?
```

Comment:
It also special-cases four dots. No effing idea why. Possibly related to the named extended attr (XATTR) support the code used to have. I think this specific test should be removed since we don't really know what the semantics are supposed to be and there shouldn't be any ksh script relying on this "feature". Instead add a TODO to the `pathcanon()` code reminding someone to remove that logic if they can do so without risking breaking other behavior.